src/java.desktop/share/classes/javax/swing/filechooser/FileSystemView.java

```diff
@@ -255,6 +256,54 @@ public Icon getSystemIcon(File f) {
         }
     }
 
+   /**
+    * Returns an icon for a file, directory, or folder as it would be displayed
+    * in a system file browser for the requested size.
+    * <p>
+    * The default implementation gets information from the
+    * {@code ShellFolder} class. Whenever possible, the icon
+    * returned is a multi-resolution icon image,
+    * which allows better support for High DPI environments
+    * with different scaling factors.
+    * <p>
+    * Example: <pre>
+    * FileSystemView fsv = FileSystemView.getFileSystemView();
+    * Icon icon = fsv.getSystemIcon("application.exe", 64);
```

Comment:
Good catch! Yes, fixed both here and in CSR.

Reply:
Are we sure that all possible paths can be pointed by the file object? Especially some "Windows Libraries" which are accessed by the shell folder?

Reply:
Later we say that this method returns null for non-existed files. is it always correct? I am not sure that the file created for the library report true for the exists() method; DId we test this usecase?

Reply:
We do not test for that in the regression test but i did tested it manually and we do return null for the non-existed files. I tested it on non-windows platform too. We still return null.

Reply:
> Are we sure that all possible paths can be pointed by the file object?

We specify that we return the icon for a file. If path can not be resolved in the file object we can not return the icon for it.

Reply:
So the libraries are not supported? I doubt since JFileChooser should support them, and supports of it is one of the reasons why we use shell folder to iterate the folders and do not use the javaio.

Reply:
No, libraries are not supported.

I see no contradiction here: `JFileChooser` uses Windows Shell API to enumerate objects and navigate the shell namespace. But it does not return non-file objects, does it?

The new method accepts `File` object which implies it's a file system object rather than an arbitrary object in Windows Shell namespace which cannot be represented in Java.

Reply:
The JFileChooser supports the libraries since it allows navigation inside them. It is done via ShellFolder which extends the file class. The FileSystemView feeds the JFileChooser by the data, so it may returns "non-file" objects.
For example the FileSystemView.getRoots() on WIndows will return "new Win32ShellFolder2(DESKTOP)".

Usually, when the FileSystemView gets a file object it checks that the file is "instance ShellFolder".
For example, in the newly added method the line "sf = ShellFolder.getShellFolder(f);" will check that the file is ShellFolder and return it as-is without checking of file existence, and really that path may not exist. but if it is a plain file it will check that.

This is why I have asked about virtual folders, do we test them or we can consider them as "not existed"?

Reply:
Maybe we can test this case by just reusing this method where the "sf.getIcon(largeicon)" previously used in the WindowsPlacesBar class? That old method has the special code for "folder.isLibrary()" so I assume it supported the Libraries. And if we will start to use the new method then all old tests for the JFileChooser will start to test it.

Reply:
What is the library a collection folder such as Recent Items or Documents?

Reply:
It is accessible from the "JFileChooser" drop-down menu. On my system, the Libraries at that drop down menu contain "GIT", "Documents", "Music". https://docs.microsoft.com/en-us/windows/client-management/windows-libraries

There are also "known folders" such as "Network" or "My computer" which do not have a real path but can be navigated in the JFileChooser and has an icon.

Reply:
> It is accessible from the "JFileChooser" drop-down menu. On my system, the Libraries at that drop down menu contain "GIT", "Documents", "Music". https://docs.microsoft.com/en-us/windows/client-management/windows-libraries
> 
> There are also "known folders" such as "Network" or "My computer" which do not have a real path but can be navigated in the JFileChooser and has an icon.

Ok, got it. Well, since they can be translated into the file system paths - even if these paths do not belong to physical filesystem - they are supported by the new API. Not for all of them i am able to receive the high resolution icons - like "Recent Items" for some reason only provides 32x32 and 16x16 no matter what size i am asking for. Others such as Documents, My Computer or 3D Objects do provide all resolutions available. For example here's the screenshot of all the available icons for 3D Objects 
![image](https://user-images.githubusercontent.com/69642324/119092517-62a9f980-b9c3-11eb-9d51-b7745cab564c.png)

Reply:
But how you got them via this method? I am not sure what parameters should be passed to it.

Reply:
Didn't you answer your question already? If `FileSystemView.getRoots()` returns Desktop in Windows shell namespace. Otherwise, I don't know a way to get a reference to a *virtual* folder in Windows shell namespace which doesn't reference a file system object.

Alex's example uses "3D Objects" folder which is one of the known folders and has its own icon instead of the standard folder icon.

It's possible that I don't understand the question clearly.

Alex's fix affects WindowsPlacesBar on the left of JFileChooser in Windows LaF, the icons there look crispier in High DPI environment.

Reply:
> But how you got them via this method? I am not sure what parameters should be passed to it.
`
String userdir = System.getenv("userprofile");
Icon icon = FileSystemView.getFileSystemView().getSystemIcon(new File(userdir + "\\3D Objects"), 64);
`

For some of the libraries getting file reference is quite easy, for some - not so much. But still doable.

Reply:
Great!